optimizely/event/event_factory.py

```diff
@@ -0,0 +1,181 @@
+# Copyright 2019 Optimizely
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+
+from .user_event import ConversionEvent, ImpressionEvent
+from .payload import Decision, EventBatch, Snapshot, SnapshotEvent, Visitor, VisitorAttribute
+from .log_event import LogEvent
+from optimizely.helpers import enums, event_tag_utils, validator
+
+CUSTOM_ATTRIBUTE_FEATURE_TYPE = 'custom'
+
+
+class EventFactory(object):
+  """ EventFactory builds LogEvent object from a given UserEvent.
+  This class serves to separate concerns between events in the SDK and the API used
+  to record the events via the Optimizely Events API ("https://developers.optimizely.com/x/events/api/index.html")
+  """
+
+  EVENT_ENDPOINT = 'https://logx.optimizely.com/v1/events'
+  HTTP_VERB = 'POST'
+  HTTP_HEADERS = {'Content-Type': 'application/json'}
+  ACTIVATE_EVENT_KEY = 'campaign_activated'
+
+  @classmethod
+  def create_log_event(cls, user_events, logger):
+    """ Create LogEvent instance.
+
+    Args:
+      user_events: An array of UserEvent instances.
+      logger: Provides a logger instance.
+
+    Returns:
+      LogEvent instance.
+    """
+
+    def _dict_clean(obj):
+      """ Helper method to remove keys from dictionary with None values. """
+
+      result = {}
+      for k, v in obj:
+        if v is None and k in ['revenue', 'value', 'tags', 'decisions']:
+          continue
+        else:
+          result[k] = v
+      return result
+
+    if not isinstance(user_events, list):
+      user_events = [user_events]
+
+    visitors = []
+
+    for user_event in user_events:
+      visitors.append(cls._create_visitor(user_event, logger))
```

Comment:
`cls._create_visitor` could return `None`. We have to filter out `None` values from the return value before appending it to `visitors`.